classes/task/merge_user_accounts.php

```diff
@@ -0,0 +1,218 @@
+            $mergerequestresult = $this->merge($mergerequestresult,
+                                                $maxattempts,
+                                                merge_request::COMPLETED_WITH_ERRORS);
+        }
+        if ($mergerequestresult->status != merge_request::COMPLETED_WITH_ERRORS &&
+            $mergerequestresult->status != merge_request::COMPLETED_WITH_SUCCESS) {
+            // Throwing exception will ensure this adhoc task is re-queued until $maxretries is reached.
+            throw new moodle_exception(get_string('failedmergerequest', 'tool_mergeusers'));
+        }
+    }
+    /**
+     * Function for merging two users.
+     */
+    private function merge(merge_request $mergerequest,
+                            int $maxattempts,
+                            int $statusforerror) {
+        $retries = $mergerequest->retries + 1;
+        $this->update_retries($mergerequest);
+        $mut = new MergeUserTool();
+        list($success, $log) = $mut->merge($mergerequest->keepuserid,
+                                            $mergerequest->removeuserid);
+        if (!$success) {
+            if ($retries >= $maxattempts) {
+                $status = merge_request::COMPLETED_WITH_ERRORS;
+                // Send a notification to administrator ?
+            } else {
+                $status = $statusforerror;
+            }
+        } else {
+            $status = merge_request::COMPLETED_WITH_SUCCESS;
+        }
+        $mergerequest->append_log($log, time());
+        if ($mergerequest->retries > $maxattempts) {
+            $log_attempts = "Reached the number of maximum attempts";
+            $mergerequest->append_log($log_attempts, time());
+            $status = merge_request::COMPLETED_WITH_ERRORS;
+        } 
+        $this->update_status_and_log($mergerequest->id,
+                                        $status,
+                                        $mergerequest->log);
+        $mergerequest->status = $status;
+        return $mergerequest;
+    }
+    /**
+     * Function for updating the status of the merging request.
+     */
+    private function update_status_and_log(int $idrecord,
+                                            int $status,
+                                            array $log): void {
+        global $DB;
+        $update = (object)[
+            'id' => $idrecord,
+            'status' => $status,
+            'log' => json_encode($log),
+        ];
+        if ($status == merge_request::COMPLETED_WITH_SUCCESS ||
+            $status == merge_request::COMPLETED_WITH_ERRORS) {
+                $update->timecompleted = time();
+        }
+        $DB->update_record(
+            merge_request::TABLE_MERGE_REQUEST,
+            $update
+        );
+    }
+    /**
+     * Function for verifying the users to keep and remove.
+     */
+    private function verify_users_to_keep_and_remove(merge_request $mergerequest): void {
```

Comment:
This is something it happens in our production environment, looking for usernames.

Imagine that on this Moodle instance, there exist the user to remove, but not the user to keep. This is a merging action very, very, very, very simple: in this example, assuming that we look for username, and only changes the username, since both users are related to the same person, the merging action is as simple as updating its username. For instance: $removeuserid = 12345, $removeuserfield = 'username', $removeuservalue = '4444444', so that this user exists. Instead, $keepuserfiel = 'username', $keepuservalue = '555555', matches no user. So, updating to username = '555555' for the given user.id = 12345 is all what we need to do.

We should consider this case, at least when the `removeuserfield` and `keepuserfield` is `'username'`. Other fields like `email` or `idnumber` may be conflicting cases, maybe, I am not sure, but it is viable for `username`s. We do this updating on our production system with a local hack, corresponding to a local cli merger that the plugin let's process merging operations.

Reply:
if you make a clear workflow we can traspose it in code. My question: how you can decide if a merge request concern a simple username update - your case - or if it is a simple error of input that return no record on keep username filter? In the last case we don't want to merge users.

Reply:
good question.

First of all, the `userfield` should be the same, even though you can set a different `removeuserfield` from `keepuserfield`. So, in this case, both `userfield` should be `username`. And the value must be not empty. Apart from this, it is the responsability from the client side to ask "meaningfull" merge requests.

Maybe the same case could apply to `idnumber` too. I do not know about the `email` field.

When the `keepuservalue` (like the username of the user to keep) does not exists on Moodle mainly it is because the person under that new `keepuservalue` has not made any login yet on Moodle, so its user is not created yet. Just this.

In our case, we double check in our cli merger that the user to keep exists on LDAP, but not in Moodle yet. We could add a setting for that on the plugin: `allow to check on LDAP if the user to keep exists`. This would bring us more security that the simple update can be done directly without the need of a full merging operation.

Reply:
I could work on the part of the LDAP if necessary. It could be a change to be added by me later on. Just to let us close this part as soon as possible during this week.